src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java

```diff
@@ -37,32 +40,56 @@
 	private final Set<String> refIds;
 	/* every toc-entry maps exactly to one block in the resulting sequence of blocks */
 	private final Map<Block,String> refIdForBlock;
+	/* every toc-entry-on-resumed maps exactly to one block in the resulting sequence of toc-entry-on-resumed blocks */
+	private final Map<Block,TocEntryOnResumedRange> rangeForResumedBlock;
 	/* mapping from block in the resulting sequence of blocks to the toc-block element that it came from */
-	private final Map<Block,Object> tocBlockForBlock;
+	private final Map<Block,TocBlock> tocBlockForBlock;
+	/* mapping from block in the resulting sequence of toc-entry-on-resumed blocks to the toc-block element that it came from */
+	private final Map<Block,TocBlock> tocBlockForResumedBlock;
 	/* parent-child relationships of toc-block elements */
-	private final Map<Object,Object> parentTocBlockForTocBlock;
+	private final Map<TocBlock,TocBlock> parentTocBlockForTocBlock;
 	/* current stack of ancestor toc-block elements */
-	private final Stack<Object> currentAncestorTocBlocks;
+	private final Stack<TocBlock> currentAncestorTocBlocks;
+    /* toc-entry-on-resumed blocks. Invariant: this.size() == tocEntryOnResumedBlocks.size() */
+    private final FormatterCoreImpl tocEntryOnResumedBlocks;
 	/* whether we are currently inside a toc-entry */
 	private boolean inTocEntry = false;
+    /* whether we are currently inside a toc-entry-on-resumed */
+    private boolean inTocEntryOnResumed = false;
+    /* optimisation for the case that there is no toc-entry-on-resumed */
+    private boolean hasTocEntryOnResumed = false;
 
 	public TableOfContentsImpl(FormatterCoreContext fc) {
 		super(fc);
 		this.refIds = new HashSet<>();
 		this.refIdForBlock = new IdentityHashMap<>();
+        this.rangeForResumedBlock = new IdentityHashMap<>();
 		this.tocBlockForBlock = new IdentityHashMap<>();
+		this.tocBlockForResumedBlock = new IdentityHashMap<>();
 		this.parentTocBlockForTocBlock = new LinkedHashMap<>();
 		this.currentAncestorTocBlocks = new Stack<>();
+        this.tocEntryOnResumedBlocks = new FormatterCoreImpl(fc);
 	}
 
 	@Override
 	public void startBlock(BlockProperties p, String blockId) {
-		Object tocBlock = new Object();
+		if (inTocEntry || inTocEntryOnResumed) {
+			throw new RuntimeException("a block cannot be started within a toc-entry or toc-entry-on-resumed");
+		}
+        
+		TocBlock tocBlock = new TocBlock();
 		if (!currentAncestorTocBlocks.isEmpty()) {
 			parentTocBlockForTocBlock.put(tocBlock, currentAncestorTocBlocks.peek());
 		}
 		currentAncestorTocBlocks.push(tocBlock);
+        
+        inTocEntry = true;  // set the context for creating a new block
```

Comment:
In any case, this changes the meaning of `inTocEntry`. The reason I added the variable was to check that no inline content is inserted outside of `toc-entry` or `toc-entry-on-resumed`. This was done in the function `assertInTocEntry`, but now that doesn't work anymore.

Reply:
After a discussion with @bertfrees I have simplified the implementation so that this "setting of the context for creating a new block" is not needed anymore.